File: framework/languages/dotnet/samples/GameQuest/Server/QuestMission/Infrastructure/ZLink/GameplayEventRouteHandler.cs
using GameQuest.QuestMission.Application;
using GameQuest.Server.Configuration;
using GameQuest.Shared;
using Zlink.Framework.Contracts.Channels;

namespace GameQuest.QuestMission.Infrastructure.ZLink;

internal sealed class GameplayEventRouteHandler(
    PlayerQuestOwnerProvisioner playerQuestOwners,
    QuestOwnerRouter ownerRouter)
    : IZLinkRouteRequestHandler<ApplyGameplayEventReq, ApplyGameplayEventRes>
{
    public async ValueTask<ApplyGameplayEventRes> HandleAsync(
        ApplyGameplayEventReq request,
        ZLinkRouteRequestContext context,
        CancellationToken cancellationToken)
    {
        if (!ownerRouter.IsLocalOwner(request.Event.PlayerId))
            return new ApplyGameplayEventRes(false);

        return await playerQuestOwners.ApplyGameplayEventAsync(request.Event, cancellationToken);
    }
}

File: framework/languages/java/samples/java/GameQuest/Server/QuestMission/src/main/java/systems/zlink/samples/gamequest/server/questmission/handlers/GameplayEventRouteHandler.java
package systems.zlink.samples.gamequest.server.questmission.handlers;

import systems.zlink.framework.channels.ZLinkRouteRequestContext;
import systems.zlink.framework.channels.ZLinkRouteRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.gamequest.server.questmission.store.QuestStore;
import systems.zlink.samples.gamequest.shared.contracts.Messages;

@ZLinkHandlerGroup("quest-owner")
public final class GameplayEventRouteHandler
    implements ZLinkRouteRequestHandler<Messages.GameplayEventEnvelope, Messages.QuestProcessingRes> {
    private final QuestStore store;

    public GameplayEventRouteHandler(QuestStore store) {
        this.store = store;
    }

    @Override
    public Messages.QuestProcessingRes handle(
        Messages.GameplayEventEnvelope request,
        ZLinkRouteRequestContext context) {
        store.markRehydrated(request.playerId());
        return store.apply(request);
    }
}
